src/utils/infra_link_selector.cpp

```diff
@@ -0,0 +1,264 @@
+        {
+            mCurrentInfraLinkDownTime = kMinTime;
+            ExitNow();
+        }
+
+        if (mCurrentInfraLinkDownTime == kMinTime)
+        {
+            mCurrentInfraLinkDownTime = now;
+        }
+
+        // Prefer `mCurrentInfraLink` if it's down for less than `kInfraLinkSelectionDelay`
+        if (now - mCurrentInfraLinkDownTime < kInfraLinkSelectionDelay)
+        {
+            std::chrono::milliseconds delay =
+                kInfraLinkSelectionDelay -
+                std::chrono::duration_cast<std::chrono::milliseconds>(now - mCurrentInfraLinkDownTime);
+
+            mTaskRunner.Post(delay, [this]() { mRequireReselect = true; });
+            ExitNow();
+        }
+    }
+
+    {
+        const char *bestInfraLink;
+        LinkState   bestState;
+
+        EvaluateInfraLinks(bestInfraLink, bestState);
+
+        // Prefer `mCurrentInfraLink` if no other infra link is up and running
+        if (mCurrentInfraLink != nullptr && bestState != kUpAndRunning)
+        {
+            ExitNow();
+        }
+
+        // Prefer the infra link in the best state
+        assert(mCurrentInfraLink != bestInfraLink);
+
+        mCurrentInfraLink         = bestInfraLink;
+        mCurrentInfraLinkDownTime = bestState == kUpAndRunning ? kMinTime : now;
+    }
+
+exit:
+    mRequireReselect = false;
+
+    if (oldInfraLink != mCurrentInfraLink)
+    {
+        if (oldInfraLink == nullptr)
+        {
+            otbrLogNotice("Infra link selected: %s", mCurrentInfraLink);
+        }
+        else
+        {
+            otbrLogWarning("Infra link switched from %s to %s", oldInfraLink, mCurrentInfraLink);
+        }
+    }
+
+    return mCurrentInfraLink;
+}
+
+void InfraLinkSelector::EvaluateInfraLinks(const char *&aBestInfraLink, LinkState &aBestInfraLinkState)
+{
+    const char *                 bestStateInfraLink = nullptr;
+    InfraLinkSelector::LinkState bestState          = kInvalid;
+
+    otbrLogInfo("Evaluating infra link among %d netifs:", mInfraLinkNames.size());
+
+    for (const char *name : mInfraLinkNames)
+    {
+        InfraLinkSelector::LinkState state = IsInfraLinkRunning(name);
+
+        otbrLogInfo("\tInfra link %s is in state %d", name, state);
+        if (bestStateInfraLink == nullptr || state > bestState)
+        {
+            bestStateInfraLink = name;
+            bestState          = state;
+        }
+    }
+
+    aBestInfraLink      = bestStateInfraLink;
+    aBestInfraLinkState = bestState;
+}
+
+InfraLinkSelector::LinkState InfraLinkSelector::IsInfraLinkRunning(const char *aInfraLinkName)
+{
+    int                          sock = 0;
+    struct ifreq                 ifReq;
+    InfraLinkSelector::LinkState state = kInvalid;
+
+    sock = SocketWithCloseExec(AF_INET6, SOCK_DGRAM, IPPROTO_IP, kSocketBlock);
+    VerifyOrDie(sock != -1, "Failed to create AF_INET6 socket.");
+
+    memset(&ifReq, 0, sizeof(ifReq));
+    strncpy(ifReq.ifr_name, aInfraLinkName, sizeof(ifReq.ifr_name));
```

Comment:
`specified bound 16 equals destination size [-Werror=stringop-truncation]`

Reply:
In your log, `wlan0` is chosen correctly because WiFi is up and running all the time.

If you prefer eth0 to wlan0 when they are both up and running, you need to specify `-B eth0 -B wlan0` (put eth0 first)

Reply:
Thanks for your response. Indeed, I realized the preference was out of the order of the arguments. However, I meant to point out that the PR doesn't support reacting to interface changes as described in the text:
> Run Infrastructure Link Selection when network interface states change is detected

Reply:
I think it reacted to the network interface change in https://github.com/openthread/ot-br-posix/pull/1353#pullrequestreview-992839960.

Reply:
That's the state after a reboot, not reacting to it while it was running.

Reply:
The current strategy is that otbr-agent will quit when infrastructure link is detected to have changed. 
We haven't implemented real run-time infrastructure switching without rebooting otbr-agent, due to its complexity and huge impact on the existing OpenThread architecture.

Reply:
`-Werror=stringop-truncation` issue is now resolved.